src/librustc_mir/monomorphize/collector.rs

```diff
@@ -714,10 +714,12 @@ fn visit_fn_use<'tcx>(
     output: &mut Vec<MonoItem<'tcx>>,
 ) {
     if let ty::FnDef(def_id, substs) = ty.kind {
-        let instance = ty::Instance::resolve(tcx,
-                                             ty::ParamEnv::reveal_all(),
-                                             def_id,
-                                             substs).unwrap();
+        let resolver = if is_direct_call {
+            ty::Instance::resolve
+        } else {
+            ty::Instance::resolve_for_fn_ptr
+        };
+        let instance = resolver(tcx, ty::ParamEnv::reveal_all(), def_id, substs).unwrap();
```

Comment:
Ah this must be where miri takes into account reifications.

Reply:
See above, I found it. Knew I must have been missing something there, I even checked earlier this morning!

Reply:
For the record: I was confused and thought this was part of miri, oops.